File: crates/brainwires-mcp-client/examples/mcp_client.rs
//! Example: MCP Client configuration and API overview
//!
//! Demonstrates how to configure `McpClient` and `McpServerConfig`,
//! and shows the full API surface for connecting to MCP servers, listing
//! tools/resources/prompts, and calling tools. Since we cannot connect to
//! a real MCP server in this example, we demonstrate the configuration and
//! API patterns using mock data.
//!
//! Run: cargo run -p brainwires-mcp --example mcp_client --features native

use brainwires_mcp_client::{McpClient, McpServerConfig};

#[tokio::main]
async fn main() {
    println!("=== MCP Client Example ===\n");

    // ── 1. Configure MCP servers ────────────────────────────────────────
    println!("--- Step 1: Server Configuration ---");

    // Define MCP server configurations
    let filesystem_server = McpServerConfig {
        name: "filesystem".to_string(),
        command: "npx".to_string(),
        args: vec![
            "-y".to_string(),
            "@modelcontextprotocol/server-filesystem".to_string(),
            "/tmp".to_string(),
        ],
        env: None,
    };
    println!("  Server: {}", filesystem_server.name);
    println!(
        "  Command: {} {}",
        filesystem_server.command,
        filesystem_server.args.join(" ")
    );
    println!();

    // Server with environment variables
    let mut env_vars = std::collections::HashMap::new();
    env_vars.insert("GITHUB_TOKEN".to_string(), "ghp_demo_token".to_string());

    let github_server = McpServerConfig {
        name: "github".to_string(),
        command: "npx".to_string(),
        args: vec![
            "-y".to_string(),
            "@modelcontextprotocol/server-github".to_string(),
        ],
        env: Some(env_vars),
    };
    println!("  Server: {}", github_server.name);
    println!(
        "  Command: {} {}",
        github_server.command,
        github_server.args.join(" ")
    );
    println!(
        "  Env vars: {:?}",
        github_server
            .env
            .as_ref()
            .map(|e| e.keys().collect::<Vec<_>>())
    );
    println!();

    // Configs serialize to JSON for persistence
    let json = serde_json::to_string_pretty(&filesystem_server).unwrap();
    println!("  Serialized config:");
    println!("  {}", json);
    println!();

    // ── 2. Create the MCP client ────────────────────────────────────────
    println!("--- Step 2: Create McpClient ---");

    let client = McpClient::new("brainwires-example", "0.1.0");
    println!("  Created McpClient");
    println!("  Connected servers: {:?}", client.list_connected().await);
    println!();

    // ── 3. Demonstrate the connection API (without real servers) ─────────
    println!("--- Step 3: Connection API Overview ---");
    println!("  The McpClient provides these async methods:\n");

    println!("  // Connect to a server (spawns process, performs MCP handshake)");
    println!("  client.connect(&server_config).await?;\n");

    println!("  // Check connection status");
    println!("  client.is_connected(\"filesystem\").await;\n");

    println!("  // List all connected servers");
    println!("  client.list_connected().await;\n");

    // ── 4. Demonstrate the tool API ─────────────────────────────────────
    println!("--- Step 4: Tool API ---");
    println!("  // List available tools from a connected server");
    println!("  let tools = client.list_tools(\"filesystem\").await?;");
    println!("  for tool in &tools {{");
    println!("      println!(\"Tool: {{:?}}\", tool);");
    println!("  }}\n");

    println!("  // Call a tool with JSON arguments");
    println!("  let result = client.call_tool(");
    println!("      \"filesystem\",");
    println!("      \"read_file\",");
    println!("      Some(serde_json::json!({{ \"path\": \"/tmp/test.txt\" }})),");
    println!("  ).await?;\n");

    // Show what a tool call looks like with mock arguments
    let mock_args = serde_json::json!({
        "path": "/tmp/example.txt",
        "encoding": "utf-8"
    });
    println!("  Mock tool arguments:");
    println!("  {}\n", serde_json::to_string_pretty(&mock_args).unwrap());

    // ── 5. Demonstrate resources and prompts API ────────────────────────
    println!("--- Step 5: Resources & Prompts API ---");
    println!("  // List resources exposed by a server");
    println!("  let resources = client.list_resources(\"filesystem\").await?;\n");

    println!("  // Read a specific resource by URI");
    println!(
        "  let content = client.read_resource(\"filesystem\", \"file:///tmp/data.json\").await?;\n"
    );

    println!("  // List prompt templates");
    println!("  let prompts = client.list_prompts(\"github\").await?;\n");

    println!("  // Get a prompt with arguments");
    println!("  let prompt = client.get_prompt(");
    println!("      \"github\",");
    println!("      \"review_pr\",");
    println!("      Some(serde_json::json!({{ \"pr_number\": 42 }})),");
    println!("  ).await?;\n");

    // ── 6. Server info and capabilities ─────────────────────────────────
    println!("--- Step 6: Server Info & Capabilities ---");
    println!("  // After connecting, query server metadata");
    println!("  let info = client.get_server_info(\"filesystem\").await?;");
    println!("  println!(\"Server: {{}} v{{}}\", info.name, info.version);\n");

    println!("  let caps = client.get_capabilities(\"filesystem\").await?;");
    println!("  // Capabilities indicate which features the server supports:");
    println!("  // - tools: server exposes callable tools");
    println!("  // - resources: server exposes readable resources");
    println!("  // - prompts: server exposes prompt templates\n");

    // ── 7. Disconnection and cancellation ───────────────────────────────
    println!("--- Step 7: Cleanup ---");
    println!("  // Disconnect from a specific server");
    println!("  client.disconnect(\"filesystem\").await?;\n");

    println!("  // Cancel a pending request (JSON-RPC $/cancelRequest)");
    println!("  client.cancel_request(\"github\", request_id).await?;\n");

    // Verify no connections remain
    println!(
        "  Final connected servers: {:?}",
        client.list_connected().await
    );

    println!("\nDone! To connect to real MCP servers, ensure the server");
    println!(
        "command is installed (e.g., `npm install -g @modelcontextprotocol/server-filesystem`)"
    );
    println!("and call client.connect(&config).await to establish a connection.");
}

File: deno/examples/mcp/mcp_client.ts
// Example: MCP Client
// Demonstrates MCP client configuration, McpServerConfig construction,
// McpConfigManager usage, and the full McpClient API surface.
// Since we cannot connect to a real MCP server in this example, we show
// configuration patterns and API overview using mock data.
// Run: deno run --allow-read --allow-write --allow-env deno/examples/mcp/mcp_client.ts

import {
  McpClient,
  McpConfigManager,
  type McpServerConfig,
} from "@rullama/mcp-client";

async function main(): Promise<void> {
  console.log("=== MCP Client Example ===\n");

  // -----------------------------------------------------------------------
  // 1. Configure MCP servers
  // -----------------------------------------------------------------------
  console.log("--- 1. Server Configuration ---");

  const filesystemServer: McpServerConfig = {
    name: "filesystem",
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-filesystem", "/tmp"],
  };
  console.log(`  Server: ${filesystemServer.name}`);
  console.log(
    `  Command: ${filesystemServer.command} ${filesystemServer.args.join(" ")}`,
  );
  console.log();

  // Server with environment variables
  const githubServer: McpServerConfig = {
    name: "github",
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-github"],
    env: { GITHUB_TOKEN: "ghp_demo_token" },
  };
  console.log(`  Server: ${githubServer.name}`);
  console.log(
    `  Command: ${githubServer.command} ${githubServer.args.join(" ")}`,
  );
  console.log(
    `  Env vars: [${Object.keys(githubServer.env ?? {}).join(", ")}]`,
  );
  console.log();

  // Configs serialize to JSON for persistence
  const json = JSON.stringify(filesystemServer, null, 2);
  console.log("  Serialized config:");
  console.log(`  ${json}`);
  console.log();

  // -----------------------------------------------------------------------
  // 2. Create the MCP client
  // -----------------------------------------------------------------------
  console.log("--- 2. Create McpClient ---");

  const client = new McpClient("rullama-example", "0.1.0");
  console.log(
    `  Created McpClient (name="${client.clientName}", version="${client.clientVersion}")`,
  );
  console.log(`  Connected servers: [${client.listConnected().join(", ")}]`);
  console.log();

  // Also show the default factory
  const defaultClient = McpClient.createDefault();
  console.log(
    `  Default McpClient: name="${defaultClient.clientName}", version="${defaultClient.clientVersion}"`,
  );
  console.log();

  // -----------------------------------------------------------------------
  // 3. Connection API overview (without real servers)
  // -----------------------------------------------------------------------
  console.log("--- 3. Connection API Overview ---");
  console.log("  The McpClient provides these methods:\n");

  console.log(
    "  // Connect to a server (spawns process, performs MCP handshake)",
  );
  console.log("  await client.connect(serverConfig);\n");

  console.log("  // Check connection status");
  console.log('  client.isConnected("filesystem");\n');

  console.log("  // List all connected servers");
  console.log("  client.listConnected();\n");
  console.log();

  // -----------------------------------------------------------------------
  // 4. Tool API
  // -----------------------------------------------------------------------
  console.log("--- 4. Tool API ---");
  console.log("  // List available tools from a connected server");
  console.log('  const tools = await client.listTools("filesystem");');
  console.log("  for (const tool of tools) {");
  console.log("    console.log(`Tool: ${tool.name} - ${tool.description}`);");
  console.log("  }\n");

  console.log("  // Call a tool with JSON arguments");
  console.log("  const result = await client.callTool(");
  console.log('    "filesystem",');
  console.log('    "read_file",');
  console.log('    { path: "/tmp/test.txt" },');
  console.log("  );\n");

  console.log("  // Call a tool with notification forwarding");
  console.log("  const result2 = await client.callToolWithNotifications(");
  console.log('    "filesystem",');
  console.log('    "read_file",');
  console.log('    { path: "/tmp/test.txt" },');
  console.log("    (notification) => console.log(notification),");
  console.log("  );\n");

  // Show mock tool arguments
  const mockArgs = {
    path: "/tmp/example.txt",
    encoding: "utf-8",
  };
  console.log("  Mock tool arguments:");
  console.log(`  ${JSON.stringify(mockArgs, null, 2)}`);
  console.log();

  // -----------------------------------------------------------------------
  // 5. Resources and prompts API
  // -----------------------------------------------------------------------
  console.log("--- 5. Resources & Prompts API ---");
  console.log("  // List resources exposed by a server");
  console.log(
    '  const resources = await client.listResources("filesystem");\n',
  );

  console.log("  // Read a specific resource by URI");
  console.log(
    '  const content = await client.readResource("filesystem", "file:///tmp/data.json");\n',
  );

  console.log("  // List prompt templates");
  console.log('  const prompts = await client.listPrompts("github");\n');

  console.log("  // Get a prompt with arguments");
  console.log("  const prompt = await client.getPrompt(");
  console.log('    "github",');
  console.log('    "review_pr",');
  console.log("    { pr_number: 42 },");
  console.log("  );\n");
  console.log();

  // -----------------------------------------------------------------------
  // 6. Server info and capabilities
  // -----------------------------------------------------------------------
  console.log("--- 6. Server Info & Capabilities ---");
  console.log("  // After connecting, query server metadata");
  console.log('  const info = client.getServerInfo("filesystem");');
  console.log("  // => { name: string, version: string }\n");

  console.log('  const caps = client.getCapabilities("filesystem");');
  console.log("  // Capabilities indicate which features the server supports:");
  console.log("  // - tools: server exposes callable tools");
  console.log("  // - resources: server exposes readable resources");
  console.log("  // - prompts: server exposes prompt templates\n");
  console.log();

  // -----------------------------------------------------------------------
  // 7. Disconnection and cancellation
  // -----------------------------------------------------------------------
  console.log("--- 7. Cleanup ---");
  console.log("  // Disconnect from a specific server");
  console.log('  await client.disconnect("filesystem");\n');

  console.log("  // Cancel a pending request (JSON-RPC $/cancelRequest)");
  console.log('  await client.cancelRequest("github", requestId);\n');

  // Verify no connections remain
  console.log(
    `  Final connected servers: [${client.listConnected().join(", ")}]`,
  );

  console.log(
    "\nDone! To connect to real MCP servers, ensure the server",
  );
  console.log(
    "command is installed (e.g., `npm install -g @modelcontextprotocol/server-filesystem`)",
  );
  console.log(
    "and call client.connect(config) to establish a connection.",
  );
}

await main();
